File: summary/go/23_beam_search.go
package main

import (
	"fmt"
	"sort"
)

type Node struct {
	sequence string
	score    float64
}

// Function to generate possible next states (dummy implementation)
func generateNextStates(node Node) []Node {
	// Here we simply add 'A' or 'B' to the current sequence
	nextStates := []Node{
		{sequence: node.sequence + "A", score: node.score + 1.0}, // Example scoring
		{sequence: node.sequence + "B", score: node.score + 0.5},
	}
	return nextStates
}

// Beam search implementation
func beamSearch(initialNode Node, beamWidth int, maxSteps int) Node {
	currentNodes := []Node{initialNode}

	for step := 0; step < maxSteps; step++ {
		allCandidates := []Node{}

		// Generate all possible next states for each node in currentNodes
		for _, node := range currentNodes {
			nextStates := generateNextStates(node)
			allCandidates = append(allCandidates, nextStates...)
		}

		// Sort candidates by score in descending order
		sort.Slice(allCandidates, func(i, j int) bool {
			return allCandidates[i].score > allCandidates[j].score
		})

		// Keep only the top beamWidth candidates
		if len(allCandidates) > beamWidth {
			allCandidates = allCandidates[:beamWidth]
		}

		// Update current nodes
		currentNodes = allCandidates
	}

	// Return the best node from the final set of nodes
	bestNode := currentNodes[0]
	for _, node := range currentNodes {
		if node.score > bestNode.score {
			bestNode = node
		}
	}
	return bestNode
}

func main() {
	initialNode := Node{sequence: "", score: 0.0}
	beamWidth := 3
	maxSteps := 5

	result := beamSearch(initialNode, beamWidth, maxSteps)
	fmt.Printf("Best sequence: %s with score: %.2f\n", result.sequence, result.score)
}

File: summary/ts/23_beam_search.ts
type State = {
    sequence: string;
    score: number;
};

function beamSearch(startState: State, expand: (state: State) => State[], beamWidth: number, goalTest: (state: State) => boolean): State | null {
    // 初期のビームを設定
    let beam: State[] = [startState];
    
    // 探索ループ
    while (beam.length > 0) {
        // 各状態を展開し、候補を収集
        let candidates: State[] = [];
        for (let state of beam) {
            if (goalTest(state)) {
                return state; // 目標状態に到達したら終了
            }
            candidates.push(...expand(state));
        }
        
        // スコアに基づいて候補をソートし、上位ビーム幅の候補を選択
        candidates.sort((a, b) => b.score - a.score); // スコアが高い順にソート
        beam = candidates.slice(0, beamWidth);
    }
    
    return null; // 目標状態に到達できなかった場合
}

// 展開関数の例
function expand(state: State): State[] {
    // 状態を展開して次の状態の配列を返す
    // この例では文字列の各文字に 'a' または 'b' を追加
    const nextStates: State[] = [];
    const nextChars = ['a', 'b'];

    for (let char of nextChars) {
        const newSequence = state.sequence + char;
        const newScore = evaluate(newSequence); // スコアリング関数
        nextStates.push({ sequence: newSequence, score: newScore });
    }

    return nextStates;
}

// スコアリング関数の例
function evaluate(sequence: string): number {
    // ここでは単純に文字列の長さをスコアとして使用
    return sequence.length;
}

// 目標状態判定関数の例
function goalTest(state: State): boolean {
    // 例えば、文字列が 'aaa' になったら目標状態とする
    return state.sequence === 'aaa';
}

// 初期状態
const startState: State = { sequence: '', score: 0 };

// ビームサーチの実行
const result = beamSearch(startState, expand, 2, goalTest);

if (result) {
    console.log(`Goal found: ${result.sequence}`);
} else {
    console.log('Goal not found');
}
